src/java.base/share/classes/jdk/internal/event/JFRTracing.java

```diff
@@ -22,14 +22,31 @@
  * or visit www.oracle.com if you need additional information or have any
  * questions.
  */
+package jdk.internal.event;
 
-package jdk.jfr.internal.instrument;
+import sun.nio.ch.FileChannelImpl;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.RandomAccessFile;
+import java.lang.Throwable;
+import java.lang.reflect.Field;
 
-import java.lang.annotation.Retention;
-import java.lang.annotation.RetentionPolicy;
+/**
+ * Helper class to enable JFR tracing.
+ */
+public final class JFRTracing {
+
+  public static void enable() throws NoSuchFieldException, IllegalAccessException {
+      enable(Throwable.class);
+      enable(FileInputStream.class);
+      enable(FileOutputStream.class);
+      enable(FileChannelImpl.class);
+      enable(RandomAccessFile.class);
+  }
 
-@Retention(RetentionPolicy.RUNTIME)
-@interface JITypeMapping {
-    String from();
-    String to();
+  private static void enable(Class<?> clazz) throws NoSuchFieldException, IllegalAccessException {
+      Field field = clazz.getDeclaredField("jfrTracing");
+      field.setAccessible(true);
+      field.setBoolean(null, true);
+  }
```

Comment:
Using reflection with `Field` seems expedient - a more modern way could be to use `VarHandle` but I guess it would require more setup to obtain a `Lookup` with the proper capabilities?

Reply:
The field is only used once and a VarHandle implementation loads three additional classes during startup and in my measurements add about 0.6 ms to startup.

Reply:
A compromise between performance and readability is:

if (JFRTracing.isEnabled()) {
      ...
}

One additional class is loaded, but it's more clear where it comes from. I didn't want to that for the ThrowableTrace since it had <clinit>.

Reply:
One thing that isn't clear (to me anyway) is how it works with the memory model. It's plain read at the use sites, looks like the set when recording is turned on is also a plain write. Would it be possible to summarise what else happens when recording is turned on?

Reply:
During JFR initialization, jfrTracing flag is set to true by the JFRTracing class. Then, when the recording starts, a safepoint occurs, and all Java threads are stopped at a poll site. When they wake up and eventually read the memory of the jfrTracing field, it will be true because of memory fences related to the safepoint.

I guess it's not 100% safe if the JIT decides to store the read value elsewhere over several event checks, but it seems unlikely. Event settings checks (i.e., Event::isEnabled()) have always used plain reads, so it is not more unreliable than any other event.

I'm fine with using a volatile too. I used it for the exception events, but I now think a plain write/read of jfrTracing is safe for all practical purposes.